src/Web/Form/Types.hs

```diff
@@ -0,0 +1,41 @@
+module Web.Form.Types where
+
+import Foundation (Handler)
+import Web.Sessions (UserAssigns)
+
+data FormValidation e a = FieldErrors (NonEmpty e)
+                        | Result a
```

Comment:
This looks suspiciously like an `Either`, the crux of the type is have the free semigroup `NonEmpty` in the left branch, which comes into play in the `Applicative` instance and allows for some nice error handling.